### ISSUE TITLE
Resolver repos are not source control friendly

### ISSUE THREAD

**Issue by bklooster** [googlesamples/unity-jar-resolver#239]
Currently, when a plugin (such as Firebase's Unity plugin) defines its own repository for the library the resolver will create an entry in the resolver path that is user specific. This makes source control very painful because it is defined to the user's file path. 
![Screen Shot 2019-06-19 at 5 40 25 PM](https://user-images.githubusercontent.com/3290316/59810392-e0d26280-92b9-11e9-9316-e7806709d099.png)

Would it be possible to manually embed only the .aar files that are in their own repository so that the mainTemplate.gradle doesn't need to be hooked up to an absolute path name? The only work around 
Currently, to get around this problem, every user needs to force resolve on their machine prior to making a build. This is is very error prone. Also as part of the resolve process these files are recreated, which causes a change to appear in git due to the .meta files changing. 
![Screen Shot 2019-06-19 at 5 49 33 PM](https://user-images.githubusercontent.com/3290316/59810639-217eab80-92bb-11e9-94e8-cce03925a43c.png)

Any recommendations or fixes in new versions would be appreciated, thanks!

**Comment by stewartmiles** [comment]
Looks like the problem is here:
https://github.com/googlesamples/unity-jar-resolver/blob/c1b77e391030a596697adaf57194160544fe0cc3/source/PlayServicesResolver/src/PlayServicesResolver.cs#L1885

Since the code injected calculates the project path it should probably be using the `unityProjectPath` variable if the repo is located in the project directory.

**Comment by stewartmiles** [comment]
Oh now I'm confused, the resolver already does this here:
https://github.com/googlesamples/unity-jar-resolver/blob/c1b77e391030a596697adaf57194160544fe0cc3/source/PlayServicesResolver/src/PlayServicesResolver.cs#L1894

**Comment by stewartmiles** [comment]
@bklooster which version are you using?

**Comment by bklooster** [comment]
I just tested this with the latest I could find 1.2.118.0 and it is still happening. 

Thank you for the code link. 

Looking at the code you have a mechanic that says only do this relative path if export project is disabled. Our build process exports a Google project, and doesn't use the built-in gradle build system because company/project specific reasons (basically, we do some post-build modifications to the studio project similar to what we do with the xcode project) which needs to have the android studio project exported.

**Comment by stewartmiles** [comment]
@bklooster how are you hooking the Android project export in Unity?  I'm wondering whether the Android Resolver could copy the local maven repos into the exported project as well that way the paths could simply be relative to the exported gradle project in all cases.

**Comment by stopiccot** [comment]
I really suggest to move `firebase-xxxx-unity` aars to maven instead of continuing to maintain these hacks. Would be so simple, straightforward  and version control friendly.

**Comment by stopiccot** [comment]
I've also noticed that running resolution creates `m2repository/com/google/firebase/firebase-app-unity/6.1.1/firebase-app-unity-6.1.1.aar` next to `m2repository/com/google/firebase/firebase-app-unity/6.1.1/firebase-app-unity-6.1.1.srcaar` and this aar copied by Unity to libs folder of exported android studio project.

**Comment by stewartmiles** [comment]
@stopiccot yes the copy to an `aar` is intentional, see https://github.com/googlesamples/unity-jar-resolver/blob/master/source/PlayServicesResolver/src/GradleTemplateResolver.cs#L87

As I mentioned we have no intention of publishing this to public Maven, instead TBH it would be nice to go the other way and support redistribution of Maven repo for all artifacts that plugins require, at the moment the build process is not friendly to hermitic builds or build servers that are locked down with limited internet access.

**Comment by stopiccot** [comment]
My point was that after resolver creates aar file next to srcaar, there is no need to add path to local maven repo and `firebase-app-unity:6.1.1` line to mainTemplate.gradle - unity will just copy aar to exported project. 

TBH we have a lot of other sdks in our project. Mainly there two approaches: 1. push everything to maven and let resolver download artifacts. 2. just place aars inside unitypackage in Plugins/Android folder. Firebase is the only SDK I know that does local maven repo thing. And as for now I'm not really sure what are the benefits of this approach. But it obviously has some drawbacks because we are having this discussion in the first place.

**Comment by stewartmiles** [comment]
@stopiccot copying the aar to the exported project is not sufficient to build with Gradle depending upon the Unity version there are different processing steps that need to be applied to aars and jars.  So we use a local maven repo to keep the source (i.e unmodified) aar / jar in a local maven repo then process them based upon the build settings, Unity version etc.

I see what you mean for the mainTemplate.gradle build though, in that case since we inject some rules to strip unused ABIs it would likely work to just copy the `srcaar` to an `aar` in the local repo.  It is a bit more messy though when examining the set of injected dependencies as the developer wouldn't see them in `mainTemplate.gradle` making it a little harder to track down what is included in the project.

As far as I'm aware there is only one issue with the current approach that is outlined by this bug.  It's currently not possible to export a Gradle project copy that to another machine or move the Unity project and build.  I'm sure there must be some way to determine the export location of the Gradle project in which case it would be possible for the resolver to stage a local maven repo in there as well.

The practice of dropping library files into projects is a very terrible thing to do in Java anyway as folks run into multiply defined symbols clashing version compatibility etc.  I would really prefer folks to either publish their libraries to Maven or place libraries in a local Maven repo that is redistributed with their plugin, then it's possible to actually check for compatibility between dependencies.

**Comment by stopiccot** [comment]
I fully agree that publishing all libraries to Maven is the best possible solution

**Comment by bklooster** [comment]
Sorry I missed your original response. Anyways, in regards to your original question you asked, we have a [PostBuild] attribute that copies values as part of the build script. We also have various steps on our automated build script that copies files from the project directory to the exported google path base on build environment, etc. We _could_ copy the files as part of this and modify the MainTemplate.gradle, but that seems fairly error prone to me. We're currently developing dozens of SKUs and maintaining compatibility with the latest/greatest firebase and resolver already is a struggle. 

Luckily our build system has external internet connectivity. I'm sure that's a big issue for some teams with the newish resolver that mangles the mainTemplate.gradle. We have encountered a situation though where the dependency repos were changed and AndroidX libs started showing up which caused us to pivot to support them sooner than we would have wanted. So having them stored locally definitely has its advantages so that we can force ourselves to not always be forced to be up-to-date. I like not having dozens of AAR files in our source control, but there are definitely some downsides to letting gradle do its thing. Including not really knowing the final resolved versions of the AARs that are being downloaded and included in the project. 

In regards to these hacky local repos, I certainly agree that it would be easiest to have them as part of a dependency chain on the public maven. The original bug still stands in that the mainTemplate.gradle mangling and local maven repo file replacing is not source-control friendly when doing an AndroidProject export.

**Comment by borogove** [comment]
> As I mentioned we have no intention of publishing this to public Maven, 

>  I would really prefer folks to either publish their libraries to Maven 

Yeah, us too.

**Comment by tallior** [comment]
any fix for this? or workaround?

**Comment by marcelo-fanatee** [comment]
a workaround i did was to force resolve before starting my build pipeline `GooglePlayServices.PlayServicesResolver.ResolveSync(true);`
not the best option but every one on my team can build

**Comment by chkuang-g** [comment]
Duplicate of #426
